File: problems/leetcode/0020-valid-parentheses/solution.go
package main

func isValid(s string) bool {
	var stack []rune

	for _, char := range s {
		var lastOpen rune
		if len(stack) > 0 {
			lastOpen = stack[len(stack)-1]
		}

		if char == '(' || char == '[' || char == '{' {
			stack = append(stack, char)
		}

		if char == ')' && lastOpen != '(' {
			return false
		} else if char == ')' && lastOpen == '(' {
			stack = stack[:len(stack)-1]
		}

		if char == ']' && lastOpen != '[' {
			return false
		} else if char == ']' && lastOpen == '[' {
			stack = stack[:len(stack)-1]
		}

		if char == '}' && lastOpen != '{' {
			return false
		} else if char == '}' && lastOpen == '{' {
			stack = stack[:len(stack)-1]
		}
	}

	return len(stack) == 0
}

File: problems/leetcode/0020-valid-parentheses/solution.ts
// GUARDAR   → push()
// MIRAR     → stack[stack.length - 1]
// ELIMINAR  → pop()
// Creo una Stack vacía. Recorro cada símbolo del string. Si el símbolo es (, lo guardo en la Stack.
function validParentheses(s: string): boolean {
  const stack: string[] = [];

  for (let i = 0; i < s.length; i++) {
    const lastOpen = stack[stack.length - 1];

    if (s[i] === "(" || s[i] === "[" || s[i] === "{") {
      stack.push(s[i]);
    }

    if (s[i] === ")" && lastOpen !== "(") {
      return false;
    } else if (s[i] === ")" && lastOpen === "(") {
      stack.pop();
    }

    if (s[i] === "]" && lastOpen !== "[") {
      return false;
    } else if (s[i] === "]" && lastOpen === "[") {
      stack.pop();
    }

    if (s[i] === "}" && lastOpen !== "{") {
      return false;
    } else if (s[i] === "}" && lastOpen === "{") {
      stack.pop();
    }
  }

  return stack.length === 0;
}

const s = "{[()]}";
console.log(validParentheses(s))
